client/internal/statemanager/manager.go

```diff
@@ -179,14 +179,18 @@ func (m *Manager) PersistState(ctx context.Context) error {
 		return nil
 	}
 
+	bs, err := json.MarshalIndent(m.states, "", "    ")
+	if err != nil {
+		return fmt.Errorf("marshal states: %w", err)
+	}
+
 	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
 	defer cancel()
 
 	done := make(chan error, 1)
-
 	start := time.Now()
 	go func() {
-		done <- util.WriteJsonWithRestrictedPermission(ctx, m.filePath, m.states)
+		done <- util.WriteBytesWithRestrictedPermission(ctx, m.filePath, bs)
```

Comment:
What will be if this function running more then 10 sec? The ticker will start a PersistState call and will be a conflict in the file.

Reply:
That's what the ctx check and deadline is for in this fn

Reply:
But the check and move is not an atomic operation. If the code runs parallel this lines with two different ctx the outcome is unpredictable.
```
	// Check context again
	if ctx.Err() != nil {
		return ctx.Err()
	}

	if err = os.Rename(tempFileName, file); err != nil {
		return fmt.Errorf("move %s to %s: %w", tempFileName, file, err)
	}
	```

Reply:
Lets do it in separated PR